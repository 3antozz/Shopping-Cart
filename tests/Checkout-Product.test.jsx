import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CheckoutProduct from "../src/Components/Checkout-Product/Checkout-Product.jsx";
import { MemoryRouter } from "react-router-dom";

describe("Product Card component", () => {
    let onChange;
    let onRemove;
    beforeEach(() => {
        onChange = vi.fn();
        onRemove = vi.fn();
        const product = {
            id: 1,
            title: "test",
            price: 69,
            category: "...",
            description: "...",
            image: "...",
            quantity: 1,
        };
        render(
            <MemoryRouter initialEntries={["/shop"]}>
                <CheckoutProduct
                    product={product}
                    onChange={onChange}
                    onRemove={onRemove}
                />
            </MemoryRouter>
        );
    });
    it("card container is rendered", () => {
        expect(screen.getByTestId(/product-container/i)).toBeInTheDocument();
    });
    it("should call prop function when clicked", async () => {
        const user = userEvent.setup();
        const cartButton = screen.getByRole("button", { name: /x/i });

        await user.click(cartButton);
        expect(onRemove).toBeCalled();
    });
    it("navigates to product link", () => {
        expect(screen.getByText(/test/i)).toHaveAttribute(
            "href",
            "/shop/products/1"
        );
    });
    it("should increment and decrement when clicked", async () => {
        const user = userEvent.setup();
        const incrementButton = screen.getByRole("button", { name: "+" });
        const decrementButton = screen.getByRole("button", { name: "-" });

        const input = screen.getByDisplayValue("1");
        await user.click(incrementButton);
        expect(input.value).toBe("2");
        await user.click(decrementButton);
        expect(input.value).toBe("1");
    });
    it("matches snapshot", () => {
        const onChange = vi.fn();
        const onRemove = vi.fn();
        const product = {
            id: 1,
            title: "test",
            price: 69,
            category: "...",
            description: "...",
            image: "...",
            quantity: 1,
        };
        const { container } = render(
            <MemoryRouter initialEntries={["/shop"]}>
                <CheckoutProduct
                    product={product}
                    onChange={onChange}
                    onRemove={onRemove}
                />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });
});
