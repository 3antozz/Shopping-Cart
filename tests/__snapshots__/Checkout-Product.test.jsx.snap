// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Card component > matches snapshot 1`] = `
<div>
  <div
    class="_productCard_824104"
    data-testid="product-container"
  >
    <a
      data-discover="true"
      href="/shop/products/1"
    >
      <img
        alt="test"
        src="..."
      />
    </a>
    <div
      class="_input_824104"
    >
      <a
        data-discover="true"
        href="/shop/products/1"
      >
        test
      </a>
      <div>
        <h3>
          Total: 
          69
           $
        </h3>
        <label
          for="quantity"
        />
        <button
          type="button"
        >
          -
        </button>
        <input
          id="quantity"
          max="100"
          min="1"
          type="number"
          value="1"
        />
        <button
          class="increment"
          type="button"
        >
          +
        </button>
      </div>
    </div>
    <button
      type="button"
    >
      X
    </button>
  </div>
</div>
`;
